### DISCUSSION TITLE
Finetuning model locally

### DISCUSSION THREAD

**Discussion by vs1785** [ggml-org/llama.cpp#8821] [Q&A]
I am learning and trying to finetune a model locally. 

I build llama.cpp but I could not find finetune or llama-finetune executable after building.

I see many references of this command in other discussions. Does it needs to be built from another repository? Or is there any command line argument to be given to make command?

**Comment by dspasyuk** [comment]
Have you tried fine tuning using  https://github.com/hiyouga/LLaMA-Factory ?

**Comment by Victoran0** [comment]
same

> I am learning and trying to finetune a model locally.
> 
> I build llama.cpp but I could not find finetune or llama-finetune executable after building.
> 
> I see many references of this command in other discussions. Does it needs to be built from another repository? Or is there any command line argument to be given to make command?

I am also having the same issue here. I am looking at using the llama-cli, llama-server or the prebuilt binary from the releases but there is no clear documentation on what to do. Have you been able to solve it?

**Comment by Victoran0** [comment]
the finetune function has been removed, see here: https://github.com/ggerganov/llama.cpp/pull/8669

**Comment by Iemand005** [comment]
Any chance for it to come back? Would be very useful I don't want to require the whole python pile to be able to finetune a model

**Reply by woof-dog** [reply]
I can promise you that the python pile will be better than the finetune experiment. The finetune binary was very simple and only good for finetuning extremely small models, like in the millions of parameters. It took hours for me to fine tune even a handful of iterations of a 1B model on my CPU using the finetune binary. On the GPU with the Python stack it happens in a matter of seconds and supports way more features germane for efficient fine-tuning, like QLoRA etc.

**Reply by Iemand005** [reply]
Indeed, but I don't want the old slow finetune back, it'd be really nice to have a new efficient native finetuning library in C++ instead of python. llama.cpp can still be expanded to support QLoRA and GPU acceleration for training, though it's unfortunately probably quite a task to implement. Though it would be a game changer if this were to be added. llama.cpp runs on a much wider range of devices and is way less bloated than pytorch and the other hundred pip dependencies of unsloth and llama-factory.
